File: dotnet/Maker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Nethereum.Util;

namespace EhterDelta.Bots.DotNet
{
    public class Maker : BaseBot
    {
        public Maker(EtherDeltaConfiguration config, ILogger logger = null) : base(config, logger)
        {

            PrintMyOrders();

            const int ordersPerSide = 1;
            var expires = Service.GetBlockNumber().Result + 10;
            var buyOrdersToPlace = ordersPerSide - Service.MyOrders.Buys.Count();
            var sellOrdersToPlace = ordersPerSide - Service.MyOrders.Sells.Count();
            var buyVolumeToPlace = EtherDeltaETH;
            var sellVolumeToPlace = EtherDeltaToken;
            var bestBuy = Service.GetBestAvailableBuy();
            var bestSell = Service.GetBestAvailableSell();

            if (bestBuy == null || bestSell == null)
            {
                Console.WriteLine("Market is not two-sided, cannot calculate mid-market");
                return;
            }

            // Make sure we have a reliable mid market
            if (Math.Abs((bestBuy.Price - bestSell.Price) / (bestBuy.Price + bestSell.Price) / 2) > 0.05m)
            {
                Console.WriteLine("Market is too wide, will not place orders");
                return;
            }

            var uc = new UnitConversion();

            var midMarket = (bestBuy.Price + bestSell.Price) / 2;
            var orders = new List<Order>();

            for (var i = 0; i < sellOrdersToPlace; i += 1)
            {
                var price = midMarket + ((i + 1) * midMarket * 0.05m);
                var amount = sellVolumeToPlace / sellOrdersToPlace;
                Console.WriteLine($"Sell { amount.ToString("N3")} @ { price:N9}");
                try
                {
                    var order = Service.CreateOrder(OrderType.Sell, expires, uc.ToWei(price), amount);
                    orders.Add(order);
                }
                catch (Exception ex)
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine(ex.Message);
                    Console.ResetColor();
                }
            }

            for (var i = 0; i < buyOrdersToPlace; i += 1)
            {
                var price = midMarket - ((i + 1) * midMarket * 0.05m);
                var amount = uc.FromWei(buyVolumeToPlace) / price / buyOrdersToPlace;
                Console.WriteLine($"Buy { amount:N3} @ { price:N9}");
                try
                {
                    var order = Service.CreateOrder(OrderType.Buy, expires, uc.ToWei(price), uc.ToWei(amount));
                    orders.Add(order);
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex);
                }
            }

            var orderTasks = new List<Task>();
            orders.ForEach(order =>
            {
                var amount = order.TokenGive == Service.ZeroToken ? order.AmountGet : order.AmountGive;
                orderTasks.Add(Service.TakeOrder(order, amount));
            });

            try
            {
                Task.WaitAll(orderTasks.ToArray());
            }
            catch (Exception ex)
            {

                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine(ex.InnerException != null ? ex.InnerException.Message : ex.Message);
                Console.ResetColor();
            }

            Console.WriteLine("Done");
        }

        void PrintMyOrders()
        {
            Console.WriteLine($"My existing buy orders: {Service.MyOrders.Buys.Count()}");
            Console.WriteLine($"My existing sell orders: {Service.MyOrders.Sells.Count()}");
        }
    }
}

File: python/maker.py
#!/usr/bin/python3
#
# maker.py: EtherDelta maker client
# =================================
#
# This client waits until the EtherDelta order book is available,
# determines a good buy and sell price and creates new buy and
# sell orders using the EtherDelta WebSocket API.
#
# Author: Tom Van Braeckel <tomvanbraeckel+etherdelta@gmail.com>
# License: MIT License (MIT)
#
# Installation:
# =============
# This client comes with service.py that contains common EtherDelta API service facilities.
# Please see service.py for instructions on how to install it and its dependencies.
#
# Configuration:
# ==============
# Copy config.ini.example to config.ini and fill in the values.
#
# Execution:
# ==========
# . venv/bin/activate   # initialize Python 3 virtual environment, not needed if you have python 3 installed system-wide
# python maker.py

import time
import sys
import configparser

# Import our Python EtherDelta API service,
# that automatically maintains the connection,
# updates the order and trade books,
# and allows for reuse of code throughout
# different EtherDelta maker and taker clients.
from etherdeltaclientservice import EtherDeltaClientService

if __name__ == "__main__":
    print("EtherDelta API client in python using websocket")
    print("More info and details in this script's source code.")
    time.sleep(1)

    # Load config
    config = configparser.ConfigParser()
    config.read('config.ini')
    userAccount = config['DEFAULT']['user_wallet_public_key']
    user_wallet_private_key = config['DEFAULT']['user_wallet_private_key']
    token = config['DEFAULT']['token_to_trade']

    es = EtherDeltaClientService()
    es.start(userAccount, token)
    print("EtherDeltaClientService started")

    es.printBalances(token, userAccount)

    while es.getBestSellOrder() == None or es.getBestBuyOrder() == None:
        print("Waiting until best sell and buy orders are known...")
        time.sleep(4)

    best_sell = es.getBestSellOrder()
    best_buy = es.getBestBuyOrder()
    best_sell_price = float(best_sell['price'])
    best_buy_price = float(best_buy['price'])
    market_width = abs((best_buy_price - best_sell_price) / ( best_buy_price + best_sell_price ) / 2.0)
    print("Market width: %.18f ETH" % market_width)
    if market_width > 0.05:
        print("ERROR: market width is too wide, will not place orders")
        sys.exit()

    midmarket = (best_buy_price + best_sell_price) / 2.0

    es.printOrderBook()
    es.printTrades()

    ordersPerSide = 1
    sellOrdersToPlace = ordersPerSide - len(es.my_orders_sells)
    sellVolumeToPlace = 1
    buyOrdersToPlace = ordersPerSide - len(es.my_orders_buys)
    buyVolumeToPlace = 1
    expires = es.getBlockNumber() + 10

    marginfactor = 0.25

    # Create sell orders
    for sellordernr in range(1,sellOrdersToPlace+1):
        price = midmarket + sellordernr * midmarket * marginfactor
        amount = sellVolumeToPlace / sellOrdersToPlace
        order = es.createOrder('sell', expires, price, amount, token, userAccount, user_wallet_private_key)
        es.send_message(order)

    # Create buy orders
    for buyordernr in range(1,buyOrdersToPlace+1):
        price = midmarket - buyordernr * midmarket * marginfactor
        amount = float(buyVolumeToPlace) / float(price) / float(buyOrdersToPlace)
        order = es.createOrder('buy', expires, price, amount, token, userAccount, user_wallet_private_key)
        es.send_message(order)

    print("Printing the user's order book every 30 seconds...")
    while True:
        es.printMyOrderBook()
        time.sleep(30)
